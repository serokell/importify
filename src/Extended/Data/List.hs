-- | This module contains additional utility functions for list.

module Extended.Data.List
       ( concatNub
       , foldMaybeMap
       , isIn
       , removeAt
       , removeAtMultiple
       ) where

import           Universum

-- | Removes element from list by given index. If there's
-- no element at such index then list returns unchanged.
removeAt :: Int -> [a] -> [a]
removeAt i l | i < 0 = l
removeAt i l = case after of
                     []       -> l
                     (_:rest) -> before ++ rest
  where (before, after) = splitAt i l

-- | Like 'removeAt' but takes list of indices to remove.
removeAtMultiple :: [Int] -> [a] -> [a]
removeAtMultiple indices = map snd
                         . filter ((`notElem` indices) . fst)
                         . zip [0..]

-- | Folds list of values with Maybe monoidal function.
foldMaybeMap :: Monoid m => (a -> Maybe m) -> [a] -> m
foldMaybeMap f = foldMap (maybeToMonoid . f)

-- | Flipped version of 'elem'.
isIn :: Eq a => [a] -> a -> Bool
isIn = flip elem
{-# INLINE isIn #-}

-- | Like 'concatMap' but also applies 'ordNub' to resulted list.
concatNub :: Ord b => (a -> [b]) -> [a] -> [b]
concatNub = ordNub ... concatMap
