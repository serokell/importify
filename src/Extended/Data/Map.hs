{-# LANGUAGE ExplicitForAll      #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies        #-}

module Extended.Data.Map
       ( groupBy
       , insertNonEmpty
       , invertMapToList
       , mapBy
       ) where

import Universum

import qualified Data.Map as Map
import qualified Data.Semigroup as Semigroup ((<>))

-- | Inserts one element to map from keys to 'NonEmpty' list.
insertNonEmpty :: Ord k => k -> v -> Map k (NonEmpty v) -> Map k (NonEmpty v)
insertNonEmpty k v = Map.insertWith (flip (Semigroup.<>)) k (one v)

-- | Groups 'NonTrivialContainer' into 'Map' by given function.
groupBy :: forall f a b .
           (NontrivialContainer f, Element f ~ a, Eq b, Ord b)
        => (a -> b)
        -> f
        -> Map b (NonEmpty a)
groupBy f = foldl' insertBy mempty
  where
    insertBy :: Map b (NonEmpty a) -> a -> Map b (NonEmpty a)
    insertBy dict value = insertNonEmpty (f value) value dict

-- | Like 'groupBy' but keeps only one element of mapping.
mapBy :: forall f a b .
         (NontrivialContainer f, Element f ~ a, Eq b, Ord b)
      => (a -> b)
      -> f
      -> Map b a
mapBy f = foldl' insertBy mempty
  where
    insertBy :: Map b a -> a -> Map b a
    insertBy dict value = Map.insert (f value) value dict

-- | Creates Map from values to keys by mapping each value to corresponding key.
invertMapToList :: forall k v . (Ord k, Ord v) => Map k [v] -> Map v k
invertMapToList = foldl' insertList mempty . Map.toList
  where
    insertList :: Map v k -> (k, [v]) -> Map v k
    insertList valueMap (key, values) =
        foldl' (\vals val -> Map.insert val key vals) valueMap values
