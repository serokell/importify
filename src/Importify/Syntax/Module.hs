-- | Syntax utilities to work with modules and their names.

module Importify.Syntax.Module
       ( getModuleTitle
       , isInsideExport
       , modulePragmas
       ) where

import Universum

import Language.Haskell.Exts (ExportSpec (EModuleContents), ExportSpecList (..), Module (..),
                              ModuleHead (..), ModuleName (..), ModulePragma)
import Language.Haskell.Names.SyntaxUtils (dropAnn, getModuleName)

{- TODO: this function used earlier, it works, but is not used anymore
   I'll keep it in case we need it again.

-- | Returns module name of the source file.
-- We can't parse the whole file to get it because default extensions are not available yet
-- so this method uses 'NonGreedy' parsing to parse only module head.
getSourceModuleName :: Text -> String
getSourceModuleName src =
    let parseResult :: ParseResult (NonGreedy (PragmasAndModuleName SrcSpanInfo))
        parseResult = parse $ toString src
        NonGreedy (PragmasAndModuleName _ _pragmas maybeModuleName) =
            fromParseResult parseResult
        ModuleName _ modNameStr =
            fromMaybe (error "File doesn't have `module' declaration") maybeModuleName
    in modNameStr
-}

-- | Get name of module by dropping annonation.
getModuleNameId :: ModuleName l -> String
getModuleNameId (ModuleName _ moduleId) = moduleId

-- | Returns name of 'Module' as a 'String'.
getModuleTitle :: Module l -> String
getModuleTitle = getModuleNameId . getModuleName

-- | Returns 'True' iff given 'ModuleName' is inside export list.
isInsideExport :: Maybe (ModuleHead l) -> ModuleName () -> Bool
isInsideExport moduleHead moduleName = moduleName `elem` exportedModules moduleHead

-- | Extracts list of module names exported with @module A@ way.
exportedModules :: Maybe (ModuleHead l) -> [ModuleName ()]
exportedModules moduleHead = qualifiedModuleNames
  where
    exports = do
        ModuleHead _ _ _ maybeExports <- moduleHead
        maybeExports

    qualifiedModuleNames :: [ModuleName ()]
    qualifiedModuleNames = do
      ExportSpecList  _ specs   <- maybe [] one exports
      EModuleContents _ eModule <- specs
      pure $ dropAnn eModule

-- | Extract all 'ModulePragma's from 'Module'.
modulePragmas :: Module l -> [ModulePragma l]
modulePragmas (Module _ _ pragmas _ _)            = pragmas
modulePragmas (XmlPage _ _ pragmas _ _ _ _)       = pragmas
modulePragmas (XmlHybrid _ _ pragmas _ _ _ _ _ _) = pragmas
