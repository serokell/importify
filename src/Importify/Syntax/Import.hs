-- | Haskell AST utilities to work with @import@ section.

module Importify.Syntax.Import
       ( getImportModuleName
       , importNamesWithTables
       , importSlice
       , importedSymbols
       , isImportImplicit
       , switchHidingImports
       ) where

import           Universum

import qualified Data.List.NonEmpty                       as NE
import           Language.Haskell.Exts                    (Extension (DisableExtension),
                                                           ImportDecl (..),
                                                           ImportSpecList (..),
                                                           KnownExtension (ImplicitPrelude),
                                                           Module (..), ModuleName,
                                                           ModuleName (..),
                                                           ModulePragma (LanguagePragma),
                                                           Name (Ident), SrcSpan (..),
                                                           SrcSpanInfo (..), combSpanInfo,
                                                           noSrcSpan, prettyExtension)
import           Language.Haskell.Names                   (NameInfo (Import, ImportPart),
                                                           Symbol)
import           Language.Haskell.Names.GlobalSymbolTable (Table)

import           Importify.Syntax.Module                  (isInsideExport)
import           Importify.Syntax.Scoped                  (InScoped, pullScopedInfo)

-- | Returns module name for 'ImportDecl' with annotation erased.
getImportModuleName :: ImportDecl l -> ModuleName ()
getImportModuleName ImportDecl{..} = () <$ importModule

-- | Returns 'True' iff import has next form:
-- @
--   import Module.Name
--   import Module.Name as Other.Name
--   import Module.Name hiding (smth)
-- @
isImportImplicit :: ImportDecl l -> Bool
isImportImplicit ImportDecl{ importQualified = True }                       = False
isImportImplicit ImportDecl{ importSpecs = Nothing }                        = True
isImportImplicit ImportDecl{ importSpecs = Just (ImportSpecList _ True _) } = True
isImportImplicit _                                                          = False

-- | This function returns name of import disregard to its
-- qualification, how this module should be referenced, i.e. like this:
-- @
--   import <whatever> A      ⇒ A
--   import <whatever> B as X ⇒ X
-- @
importReferenceName :: ImportDecl l -> ModuleName ()
importReferenceName ImportDecl{ importAs = Nothing, .. } = () <$ importModule
importReferenceName ImportDecl{ importAs = Just name   } = () <$ name

-- | Keep only hiding imports making them non-hiding. This function
-- needed to collect unused hiding imports because @importTable@ doesn't
-- track information about hiding imports.
switchHidingImports :: Module SrcSpanInfo -> Module SrcSpanInfo
switchHidingImports (Module ml mhead mpragmas mimports mdecls) =
    Module ml
           mhead
           ( LanguagePragma noSrcSpan
                            [ Ident noSrcSpan
                            $ prettyExtension
                            $ DisableExtension ImplicitPrelude
                            ]
           : mpragmas)
           (mapMaybe unhide mimports)
           mdecls
  where
    unhide :: ImportDecl SrcSpanInfo -> Maybe (ImportDecl SrcSpanInfo)
    unhide decl = do
        ImportSpecList l isHiding imports <- importSpecs decl
        guard isHiding
        guard $ not $ isInsideExport mhead (importReferenceName decl)
        pure $ decl { importSpecs = Just $ ImportSpecList l False imports }

switchHidingImports m = m

-- | Collect mapping from import name to to list of symbols it exports.
importNamesWithTables :: [InScoped ImportDecl] -> [(ModuleName (), Table)]
importNamesWithTables = map (getImportModuleName &&& getImportedSymbols)
  where
    getImportedSymbols :: InScoped ImportDecl -> Table
    getImportedSymbols = fromImportInfo . pullScopedInfo

    fromImportInfo :: NameInfo l -> Table
    fromImportInfo (Import dict) = dict
    fromImportInfo _             = mempty

-- | Extracts list of imported symbols from 'ImportPart'. Should be
-- called for implicit imports.
importedSymbols :: InScoped ImportDecl -> [Symbol]
importedSymbols ImportDecl{..} = do
    ImportPart symbols <- [pullScopedInfo importModule]
    symbols

-- | Returns pair of line numbers — first and last line of import section
-- if any import is in list.
importSlice :: [ImportDecl SrcSpanInfo] -> Maybe (Int, Int)
importSlice []               = Nothing
importSlice [ImportDecl{..}] = Just $ startAndEndLines importAnn
importSlice (x:y:xs)         = Just $ startAndEndLines
                                    $ combSpanInfo (importAnn x) (importAnn $ NE.last (y :| xs))

startAndEndLines :: SrcSpanInfo -> (Int, Int)
startAndEndLines (SrcSpanInfo SrcSpan{..} _) = (srcSpanStartLine, srcSpanEndLine)
