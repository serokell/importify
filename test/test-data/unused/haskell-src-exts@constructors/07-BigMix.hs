module BigMix where

import Data.Bool (not, otherwise, (&&), (||))
import Data.List (sort, sortOn)
import qualified Data.List                    as L
import           Data.Monoid                  (Monoid)
import Language.Haskell.Exts (Module (Module), ModuleName, parseFile)
import qualified Language.Haskell.Exts        as HSE

import Language.Haskell.Exts.Syntax (Assoc, Decl (ForExp, FunBind, TypeSig),
                                     Namespace)

foo :: Language.Haskell.Exts.Module ()
foo = undefined
foo2 = not
foo3 = sortOn
baz = TypeSig
