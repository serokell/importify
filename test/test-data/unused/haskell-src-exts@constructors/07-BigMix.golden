module BigMix where

import Data.Bool (not)
import Data.List (      sortOn)
import Language.Haskell.Exts (Module         )

import Language.Haskell.Exts.Syntax (       Decl (                 TypeSig))

foo :: Language.Haskell.Exts.Module ()
foo = undefined
foo2 = not
foo3 = sortOn
baz = TypeSig
